example/Binary.hs

```diff
@@ -0,0 +1,389 @@
+              subo q2mr n2 rr
+              splito rr r r1 (inject' [])
+              divo n1 m q1 r1)
+          & enter')
+    ]
+{- FOURMOLU_ENABLE -}
+
+splito :: Term Binary -> Term Binary -> Term Binary -> Term Binary -> Goal ()
+splito n r n1 n2 =
+  disjMany
+    [ do
+        n === inject' []
+        n1 === inject' []
+        n2 === inject' []
+    , do
+        (b, n') <- cons n
+        b === inject' O
+        poso n'
+
+        r === inject' []
+        n1 === n'
+        n2 === inject' []
+    , do
+        (b, n') <- cons n
+        b === inject' I
+
+        r === inject' []
+        n1 === n'
+        n2 === inject' [I]
+    , do
+        (b, n') <- cons n
+        b === inject' O
+        poso n'
+
+        (_, r') <- cons r
+        n2 === inject' []
+        splito n' r' n1 n2
+    , do
+        (b, n') <- cons n
+        b === inject' I
+
+        (_, r') <- cons r
+        n2 === inject' [I]
+        splito n' r' n1 (inject' [])
+    , do
+        (b, n') <- cons n
+        (_, r') <- cons r
+
+        n2' <- fresh
+        poso n2'
+        n2 === Value (LogicCons b n2')
+        splito n' r' n1 n2'
+    ]
+
+trimap :: (a -> b) -> (a, a, a) -> (b, b, b)
+trimap f (x, y, z) = (f x, f y, f z)
+
+quadmap :: (a -> b) -> (a, a, a, a) -> (b, b, b, b)
+quadmap f (x, y, z, w) = (f x, f y, f z, f w)
+
+tryExtract' :: (Logical a) => Term a -> Either (Term a) a
+tryExtract' x = maybe (Left x) Right (extract' x)
+
+example :: IO ()
+example = do
+  putStrLn "addo 3 5 x:"
+  mapM_ print $ extract' <$> run (addo (inject' 3) (inject' 5))
+  putStrLn "\naddo 2 x 8:"
+  mapM_ print $ extract' <$> run (\x -> addo (inject' 2) x (inject' 8))
+  putStrLn "\naddo 8 x 2:"
+  print $ extract' <$> run (\x -> addo (inject' 8) x (inject' 2))
+  putStrLn "\naddo x y 8:"
+  mapM_ print $ bimap extract' extract' <$> run (\(x, y) -> addo x y (inject' 8))
+  putStrLn "\naddo x y z:"
+  mapM_ print $ take 10 $ trimap tryExtract' <$> run (\(x, y, z) -> addo x y z)
+
+  putStrLn "\nmulo 3 4 x:"
+  mapM_ print $ extract' <$> run (mulo (inject' 3) (inject' 4))
+  putStrLn "\nmulo 2 x 12:"
+  mapM_ print $ extract' <$> run (\x -> mulo (inject' 2) x (inject' 12))
+  putStrLn "\nmulo 12 x 2:"
+  print $ extract' <$> run (\x -> mulo (inject' 12) x (inject' 2))
+  putStrLn "\nmulo x y 12:"
+  mapM_ print $ bimap extract' extract' <$> run (\(x, y) -> mulo x y (inject' 12))
+  putStrLn "\nmulo x y z:"
+  mapM_ print $ take 10 $ trimap tryExtract' <$> run (\(x, y, z) -> mulo x y z)
+
+  putStrLn "\ndivo 15 4 q r:"
+  mapM_ print $ bimap extract' extract' <$> run (uncurry (divo (inject' 15) (inject' 4)))
+  putStrLn "\ndivo n 3 2 1:"
+  mapM_ print $ extract' <$> run (\n -> divo n (inject' 3) (inject' 2) (inject' 1))
+  putStrLn "\ndivo 13 m 2 1:"
+  mapM_ print $ extract' <$> run (\m -> divo (inject' 13) m (inject' 2) (inject' 1))
+  putStrLn "\ndivo n m q r:"
+  mapM_ print $ take 10 $ quadmap tryExtract' <$> run (\(n, m, q, r) -> divo n m q r)
```

Comment:
Perhaps, instead of `example` these should be doctests for corresponding functions?

Reply:
Moved most of examples to doctests, but left a few examples in `example` since this module is part of the example application.